velox/type/DecimalUtil.h

```diff
@@ -193,6 +194,50 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  /// Rescale a double value to decimal value.
+  ///
+  /// Use `folly::tryTo` to convert a double value to int128_t or int64_t. It
+  /// returns an error when overflow occurs so that we could determine whether
+  /// an overflow occurs through checking the result.
+  ///
+  /// Normally, return the rescaled value. Otherwise, if the `toValue` overflows
+  /// the TOutput's limits or the `toValue` exceeds the precision's limits, it
+  /// will throw an exception.
+  template <typename TOutput>
+  inline static TOutput
+  rescaleDouble(double inputValue, const int toPrecision, const int toScale) {
+    VELOX_USER_CHECK(
```

Comment:
https://github.com/facebookincubator/velox/pull/5307/files#diff-c6d2f99b68772f908f2f3e8b979b3c2894f790f20af68684e7b0400bed969389R212

I think the old comment means for the `try_cast`, we should not throw exception because it is expensive, just return the error, and make sure the error message is same with Presto.

Reply:
It was said in this comment https://github.com/facebookincubator/velox/pull/5767#discussion_r1284244922, that rescaleDouble's return value could be `TOutput`. And there probably is an infinity value in `inputValue`, so we must check here. This is not an expensive exception, it is a needed exception. If there is an infinity value, we could fail fast. The presto does the same logic.

Reply:
We can use `setError` to set the exception in context, not throw the exception and try catch it in `applyToSelectedNoThrowLocal`. In C++, the try catch is expensive comparing to return value. please consider this discussion. https://github.com/facebookincubator/velox/pull/5307#discussion_r1301595745

Reply:
I believe that #5801 is optimized for the `try_cast` function in presto. But for the `cast` function, we still need to throw an exception if we can't convert an input value to a target type. And I think that's the reason why #5801 didn't remove `VELOX_USER_CHECK` or `VELOX_USER_FAIL` in `DecimalUtils.h`. So probably I should modify `CastExpr::applyDoubleToDecimal` in `CastExpr-impl.h` instead of here. I am not sure, so could you tell me what's your opinion?

Reply:
try_cast is implemented using cast by turning failures into nulls. Hence, we should avoid throwing exceptions in cast and rather use setError API to report errors.